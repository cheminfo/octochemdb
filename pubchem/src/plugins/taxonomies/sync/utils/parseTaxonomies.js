import Debug from '../../../../utils/Debug.js';

const debug = Debug('parseTaxonomies');

export function* parseTaxonomies(arrayBuffer, nodes, connection) {
  try {
    const decoder = new TextDecoder();
    arrayBuffer = new Uint8Array(arrayBuffer);
    let begin = 0;
    let end = 0;
    while (end < arrayBuffer.length) {
      if (arrayBuffer[end] === 10) {
        const line = decoder.decode(arrayBuffer.subarray(begin, end));

        const fields = line.split(/[\t ]*\|[ \t]*/);
        if (fields.length < 2) {
          end++;
          continue;
        }

        let taxonomy = {};
        if (fields[9] !== '') {
          taxonomy.superkingdom = fields[9];
        }
        if (fields[8] !== '') {
          taxonomy.kingdom = fields[8];
        }
        if (fields[7] !== '') {
          taxonomy.phylum = fields[7];
        }
        if (fields[6] !== '') {
          taxonomy.class = fields[6];
        }
        if (fields[5] !== '') {
          taxonomy.order = fields[5];
        }
        if (fields[4] !== '') {
          taxonomy.family = fields[4];
        }
        if (fields[3] !== '') {
          taxonomy.genus = fields[3];
        }
        if (fields[2] !== '') {
          taxonomy.species = fields[2];
        }
        if (fields[1] !== '') {
          let taxId = Number(fields[0]);
          let rank = nodes[taxId];

          if (
            [
              'superkingdom',
              'species',
              'genus',
              'class',
              'phylum',
              'family',
              'kingdom',
            ].includes(rank)
          ) {
            taxonomy[rank] = fields[1];
          }
        }
        const entry = {
          _id: Number(fields[0].replace(/[\r\n]/g, '')),
          data: taxonomy,
        };

        yield entry;
        begin = end;
      }
      end++;
    }
    return;
  } catch (e) {
    if (connection) {
      debug(e, { collection: 'taxonomies', connection });
    }
  }
}
