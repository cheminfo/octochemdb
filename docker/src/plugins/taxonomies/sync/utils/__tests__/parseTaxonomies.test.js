import { readFileSync } from 'fs';
import { join } from 'path';

import { test, expect } from 'vitest';

import { parseTaxonomies } from '../parseTaxonomies';

test(
  'taxonomyParser',
  () => {
    const results = [];
    const arrayBuffer = readFileSync(
      join(__dirname, 'data/rankedLineageTest.dmp'),
    );
    const nodes = readFileSync(join(__dirname, 'data/nodesTest.dmp'));
    let i = 0;
    for (const entry of parseTaxonomies(arrayBuffer, nodes)) {
      if (i > 40) {
        break;
      }
      i++;
      results.push(entry);
    }
    expect(results).toMatchSnapshot();
  },
  { timeout: 30000 },
);
