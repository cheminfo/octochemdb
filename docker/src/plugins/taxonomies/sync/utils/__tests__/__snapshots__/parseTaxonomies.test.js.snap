// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`taxonomyParser 1`] = `
[
  {
    "_id": 1,
    "data": {},
  },
  {
    "_id": 131567,
    "data": {},
  },
  {
    "_id": 2157,
    "data": {},
  },
  {
    "_id": 1935183,
    "data": {
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798909,
    "data": {
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798916,
    "data": {
      "phylum": "Candidatus Baldrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798922,
    "data": {
      "class": "Candidatus Baldrarchaeia",
      "phylum": "Candidatus Baldrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798928,
    "data": {
      "class": "Candidatus Baldrarchaeia",
      "order": "Candidatus Baldrarchaeales",
      "phylum": "Candidatus Baldrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798934,
    "data": {
      "class": "Candidatus Baldrarchaeia",
      "family": "Candidatus Baldrarchaeaceae",
      "order": "Candidatus Baldrarchaeales",
      "phylum": "Candidatus Baldrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798940,
    "data": {
      "class": "Candidatus Baldrarchaeia",
      "family": "Candidatus Baldrarchaeaceae",
      "genus": "Candidatus Baldrarchaeum",
      "order": "Candidatus Baldrarchaeales",
      "phylum": "Candidatus Baldrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2799004,
    "data": {
      "phylum": "Candidatus Baldrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2799005,
    "data": {
      "phylum": "Candidatus Baldrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798908,
    "data": {
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798915,
    "data": {
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798921,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798927,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798933,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "family": "Candidatus Borrarchaeaceae",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2848170,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "family": "Candidatus Borrarchaeaceae",
      "genus": "Candidatus Borrarchaeum",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2798939,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "family": "Candidatus Borrarchaeaceae",
      "genus": "Candidatus Borrarchaeum",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2841639,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "family": "Candidatus Borrarchaeaceae",
      "genus": "Candidatus Borrarchaeum",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2841640,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "family": "Candidatus Borrarchaeaceae",
      "genus": "Candidatus Borrarchaeum",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2846742,
    "data": {
      "class": "Candidatus Borrarchaeia",
      "family": "Candidatus Borrarchaeaceae",
      "genus": "Candidatus Borrarchaeum",
      "order": "Candidatus Borrarchaeales",
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2799002,
    "data": {
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2799003,
    "data": {
      "phylum": "Candidatus Borrarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2827219,
    "data": {
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2830915,
    "data": {
      "phylum": "Candidatus Freyarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2830916,
    "data": {
      "phylum": "Candidatus Freyarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2795490,
    "data": {
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2799006,
    "data": {
      "phylum": "Candidatus Gerdarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2795491,
    "data": {
      "phylum": "Candidatus Gerdarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 1936272,
    "data": {
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2026747,
    "data": {
      "phylum": "Candidatus Heimdallarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 1841596,
    "data": {
      "phylum": "Candidatus Heimdallarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2012493,
    "data": {
      "phylum": "Candidatus Heimdallarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2608077,
    "data": {
      "phylum": "Candidatus Heimdallarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 1841597,
    "data": {
      "phylum": "Candidatus Heimdallarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 1841598,
    "data": {
      "phylum": "Candidatus Heimdallarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2876573,
    "data": {
      "phylum": "Candidatus Heimdallarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2876572,
    "data": {
      "phylum": "Candidatus Heimdallarchaeota",
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2572044,
    "data": {
      "superkingdom": "Archaea",
    },
  },
  {
    "_id": 2719382,
    "data": {
      "phylum": "Candidatus Helarchaeota",
      "superkingdom": "Archaea",
    },
  },
]
`;
